src/mempool/block_receiver.rs

```diff
@@ -0,0 +1,190 @@
+                }
+
+            }
+        }
+    }
+
+    async fn submit_block(&mut self, block: &Block) {
+        if self.validate_block_events(&block) {
+            for event in block.events.iter() {
+                info!(
+                    shard = self.shard_id.to_string(),
+                    seqnum = event.data.as_ref().unwrap().seqnum.to_string(),
+                    "Submitting block event to mempool"
+                );
+                self.mempool_tx
+                    .send(MempoolRequest::AddMessage(
+                        MempoolMessage::BlockEvent {
+                            for_shard: self.shard_id,
+                            message: event.clone(),
+                        },
+                        MempoolSource::Local,
+                        None,
+                    ))
+                    .await
+                    .unwrap();
+            }
+        }
+    }
+
+    async fn sync_missing_block_events(&mut self, start_seqnum: u64, stop_seqnum: u64) {
+        info!(start_seqnum, stop_seqnum, "Syncing missing blocks",);
+        let mut currrent_seqnum = start_seqnum;
+        while currrent_seqnum <= stop_seqnum {
+            let (block_tx, block_rx) = oneshot::channel::<Option<Block>>();
+            self.system_tx
+                .send(SystemMessage::BlockRequest {
+                    block_event_seqnum: currrent_seqnum,
+                    block_tx,
+                })
+                .await
+                .unwrap();
+            let block = block_rx.await.unwrap().unwrap();
+            self.submit_block(&block).await;
+            if let Some(last_event) = block.events.last() {
+                currrent_seqnum = last_event.seqnum() + 1;
+            }
+        }
+    }
+
+    pub async fn run(&mut self) {
+        info!(shard = self.shard_id.to_string(), "Running block receiver");
+        loop {
+            let block = self.block_rx.recv().await.unwrap();
+            info!(
+                shard = self.shard_id.to_string(),
+                num_events = block.events.len(),
+                height = block.header.as_ref().unwrap().height.unwrap().block_number,
+                "Received block"
+            );
+            if block.events.len() == 0 {
+                continue;
+            }
+            // The db is the source of truth, it's possible to read this out of the events_rx channel but delivery over that channel is not reliable (it's a broadcast channel) we may not have the most up to date state.
+            let last_stored_event_seqnum = self.stores.block_event_store.max_seqnum().unwrap();
+            let last_event_in_block = block.events.last().unwrap();
+            if last_event_in_block.seqnum() < last_stored_event_seqnum {
+                continue;
+            }
+
+            let first_event_in_block = block.events.first().unwrap();
+            if first_event_in_block.seqnum() > last_stored_event_seqnum + 1 {
+                self.sync_missing_block_events(
```

Comment:
We should add some batching here so that initial sync or even later sync if big doesn't throw a bunch of messages in the mempool at the same time. It's also hard to set a reasonable confirmation timeout without some batching.